src/subdomains/core/fee/fee.service.ts

```diff
@@ -0,0 +1,230 @@
+      throw new BadRequestException('Max usages for discount code taken');
+
+    await this.userDataService.addDiscountCode(userData, fee.id.toString());
+  }
+
+  async getFeeByDiscountCode(discountCode: string): Promise<Fee> {
+    const fee = await this.feeRepo.findOneBy({ discountCode });
+    if (!fee) throw new NotFoundException(`Discount code ${discountCode} not found`);
+    return fee;
+  }
+
+  async getUserFee(request: UserFeeRequest): Promise<number> {
+    const userFees = await this.getValidUserFees(request);
+
+    return this.calculateFee(userFees, request.userData.id);
+  }
+
+  async getDefaultFee(request: FeeRequestBase, accountType = AccountType.PERSONAL): Promise<number> {
+    const baseFees = await this.getBaseFees({ ...request, accountType });
+    const discounts = await this.getFreeDiscounts({ ...request, accountType });
+
+    return this.calculateFee([...baseFees, ...discounts]);
+  }
+
+  // --- HELPER METHODS --- //
+
+  private calculateFee(fees: Fee[], userDataId?: number): number {
+    let userBaseFee = undefined;
+    let userCustomFee = undefined;
+    let userDiscount = 0;
+
+    for (const fee of fees) {
+      switch (fee.type) {
+        case FeeType.BASE:
+          if (!userBaseFee || userBaseFee > fee.value) userBaseFee = fee.value;
+          break;
+        case FeeType.CUSTOM:
+          if (!userCustomFee || userCustomFee < fee.value) userCustomFee = fee.value;
+          break;
+        case FeeType.DISCOUNT:
+          if (userDiscount < fee.value) userDiscount = fee.value;
+          break;
+      }
+    }
+
+    if (!userBaseFee) throw new InternalServerErrorException('Base Fee is missing');
+    if (userBaseFee - userDiscount < 0) {
+      this.logger.warn(`UserDiscount higher userBaseFee! UserDataId: ${userDataId}`);
+      userDiscount = 0;
+    }
+
+    return userCustomFee ?? userBaseFee - userDiscount;
+  }
+
+  private async getValidUserFees(request: UserFeeRequest): Promise<Fee[]> {
+    const userFees: Fee[] = [];
+    const accountType = request.userData.accountType;
+
+    userFees.push(
+      ...(await this.getBaseFees({ ...request, accountType })),
+      ...(await this.getFreeDiscounts({ ...request, accountType })),
+    );
+
+    const discountCodes = request.userData.discounts?.split(';') ?? [];
+
+    for (const feeId of discountCodes) {
+      const fee = await this.feeRepo.findOneBy({ id: +feeId });
+
+      if (this.isExpiredFee(fee, { ...request, accountType })) {
+        await this.userDataService.removeDiscountCode(request.userData, fee.id.toString());
+        continue;
+      }
+
+      if (!this.isValidFee(fee, { ...request, accountType })) continue;
+
+      userFees.push(fee);
+    }
+
+    return userFees.sort((a, b) => {
+      const TypeOrder: { [b in FeeType]: number } = {
+        [FeeType.CUSTOM]: 1,
+        [FeeType.BASE]: 2,
+        [FeeType.DISCOUNT]: 3,
+      };
+      return TypeOrder[a.type] - TypeOrder[b.type];
+    });
+  }
```

Comment:
I would try to 
1. get all matching fees in one single call (e.g. filtering by accountType shouldn't be necessary, this is done in `isValidFee`), incl. getting all the discounts,
2. find all expired discount fees (in this array) and remove the fee from UserData
3. filter all matching fees with the `isValidFee` method